app/shared/app-data/src/commonMain/kotlin/domain/media/cache/storage/MediaCacheMigrator.kt

```diff
@@ -0,0 +1,233 @@
+ *
+ * https://github.com/open-ani/ani/blob/main/LICENSE
+ */
+
+package me.him188.ani.app.domain.media.cache.storage
+
+import androidx.datastore.core.DataStore
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import me.him188.ani.app.data.repository.user.SettingsRepository
+import me.him188.ani.app.domain.media.cache.MediaCacheManager
+import me.him188.ani.app.domain.media.cache.engine.HttpMediaCacheEngine
+import me.him188.ani.app.domain.media.cache.engine.TorrentMediaCacheEngine
+import me.him188.ani.app.platform.AppTerminator
+import me.him188.ani.app.platform.ContextMP
+import me.him188.ani.app.platform.files
+import me.him188.ani.utils.coroutines.IO_
+import me.him188.ani.utils.httpdownloader.DownloadState
+import me.him188.ani.utils.io.SystemPath
+import me.him188.ani.utils.io.absolutePath
+import me.him188.ani.utils.io.actualSize
+import me.him188.ani.utils.io.exists
+import me.him188.ani.utils.io.isDirectory
+import me.him188.ani.utils.io.moveDirectoryRecursively
+import me.him188.ani.utils.io.name
+import me.him188.ani.utils.io.resolve
+import me.him188.ani.utils.logging.error
+import me.him188.ani.utils.logging.info
+import me.him188.ani.utils.logging.logger
+import org.koin.core.component.KoinComponent
+
+/**
+ * Since 4.9, Default directory of torrent cache is changed to external/shared storage and
+ * cannot be changed. This is the workaround for startup migration.
+ *
+ * Since 4.11, Default directory of web m3u cache is changed to external/shared storage and
+ * cannot be changed. This is the workaround for startup migration.
+ *
+ * This class should be called only `AniApplication.Instance.requiresTorrentCacheMigration` is true,
+ * which means we are going to migrate torrent caches from internal storage to shared/external storage.
+ */
+class MediaCacheMigrator(
+    private val context: ContextMP,
+    private val metadataStore: DataStore<List<MediaCacheSave>>,
+    private val m3u8DownloaderStore: DataStore<List<DownloadState>>,
+    private val mediaCacheManager: MediaCacheManager,
+    private val settingsRepo: SettingsRepository,
+    private val appTerminator: AppTerminator,
+    private val migrateTorrent: StateFlow<Boolean>,
+    private val migrateWebM3u: StateFlow<Boolean>,
+
+    private val getNewBaseSaveDir: suspend () -> SystemPath?,
+    private val getPrevTorrentSaveDir: suspend () -> SystemPath
+) : KoinComponent {
+    private val logger = logger<MediaCacheMigrator>()
+
+    private val _status: MutableStateFlow<Status?> = MutableStateFlow(null)
+    val status: StateFlow<Status?> = _status
+
+    @OptIn(DelicateCoroutinesApi::class)
+    fun migrate() = GlobalScope.launch(Dispatchers.IO_) {
```

Comment:
```suggestion
    fun startMigrationInBackground() = GlobalScope.launch(Dispatchers.IO_) {
```


`migrate` 听起来是 either blocking or suspend

Reply:
It is actually blocking user's interactive.

Reply:
这个函数是 launch, 跟用户没关系